flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+		webRefreshIntervalMillis = config.getLong(HistoryServerOptions.HISTORY_SERVER_WEB_REFRESH_INTERVAL);
+
+		String webDirectory = config.getString(HistoryServerOptions.HISTORY_SERVER_WEB_DIR);
+		if (webDirectory == null) {
+			webDirectory = System.getProperty("java.io.tmpdir") + "flink-web-history-" + UUID.randomUUID();
+		}
+		webDir = new File(webDirectory);
+
+		String refreshDirectories = config.getString(HistoryServerOptions.HISTORY_SERVER_DIR);
+		if (refreshDirectories == null) {
+			throw new IOException(HistoryServerOptions.HISTORY_SERVER_DIR + " was not configured.");
+		}
+		for (String refreshDirectory : refreshDirectories.split(",")) {
+			refreshDirs.add(new Path(refreshDirectory));
+		}
+		refreshIntervalMillis = config.getLong(HistoryServerOptions.HISTORY_SERVER_REFRESH_INTERVAL);
+
+		this.fetcherTask = new JobArchiveFetcherTask(refreshDirs, cachedArchives, webDir);
+
+		// add shutdown hook for deleting the directories and remaining temp files on shutdown
+		try {
+			Runtime.getRuntime().addShutdownHook(new Thread() {
+				@Override
+				public void run() {
+					HistoryServer.this.stop();
+				}
+			});
+		} catch (IllegalStateException e) {
+			// race, JVM is in shutdown already, we can safely ignore this
+			LOG.debug("Unable to add shutdown hook, shutdown already in progress", e);
+		} catch (Throwable t) {
+			// these errors usually happen when the shutdown is already in progress
+			LOG.warn("Error while adding shutdown hook", t);
+		}
+	}
+
+	public void run() {
+		try {
+			start();
+			while (true) {
+				try {
+					new CountDownLatch(1).await();
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}
+			}
+		} catch (Exception e) {
+			LOG.error("Failure while running HistoryServer.", e);
+		} finally {
+			stop();
+		}
+	}
+
+	// =================================================================================================================
+	// Life-cycle
+	// =================================================================================================================
+	private void start() throws IOException, InterruptedException {
+		synchronized (startupShutdownLock) {
+			LOG.info("Starting history server.");
+
+			Files.createDirectories(webDir.toPath());
+			LOG.info("Using directory {} as local cache.", webDir);
+			if (LOG.isInfoEnabled()) {
+				for (Path refreshDir : refreshDirs) {
+					LOG.info("Monitoring directory {} for archived jobs.", refreshDir);
+				}
+			}
+
+			Router router = new Router();
+			router.GET("/:*", new HistoryServerStaticFileServerHandler(webDir, new JobFileFetcher(new Path(webDir.getAbsolutePath()), cachedArchives)));
+
+			netty = new NettySetup(router, LOG, webDir, serverSSLContext, webAddress, webPort);
+
+			createDashboardConfigFile();
+			updateJobOverview(webDir);
+
+			executor.scheduleWithFixedDelay(fetcherTask, refreshIntervalMillis, refreshIntervalMillis, TimeUnit.MILLISECONDS);
```

Comment:
The first execution could be immediate in order to not have a high startup delay until you see historic jobs.